### PR TITLE
Add query_cache.py to speed repeated queries

### LINKED ISSUE #380: Support caching of results
**I'm submitting a ...**
- [ ] bug report
- [x] feature request

Hey, currently, whenever a request comes in, query-server goes to the search engine every time to fetch the result. We can improve this, two ways
- Use something in python for caching
- Use mongodb to save results

I personally don't feel first option is good since, we'll need the results for quite a long time, so probably a lot of ram will be consumed with simple chaching, so I recommend mongodb.

There is already a basic [implementation](https://github.com/fossasia/query-server/blob/master/app/server.py#L65-L67)

### PR DESCRIPTION
**Pull request by cclauss**
Fixes #380, #390

#### Checklist

- [x] I have read the [Contribution & Best practices Guide](https://blog.fossasia.org/open-source-developer-guide-and-best-practices-at-fossasia) and my PR follows them.
- [x] My branch is up-to-date with the Upstream `master` branch.
- [x] I have added necessary documentation (if appropriate)

#### Changes proposed in this pull request:
- Create a new query_cache.py file to isolate all logic for caching queries.
- lookup() function to return any cached results that are less than one week old.
- store() function to put new search results into the cache

query_cache.py -- Implements a caching system for query server based on MongoDB
 
Before sending a query to a remote search engine, use lookup() see if results from that same search engine and query are already in the cache.  If so, then return the cached results.  If not, then use store() to write the search engine, query, query results, and a datetime created into the cache.  MongoDB uses datetime will be used to automatically delete out dated query results.

Ideas for improvement:
* Add a lookup_count to see how often cache actually saves us time.

### CODE REVIEW

File: app/server.py [hunk 1]
```diff
@@ -59,14 +60,17 @@ def search(search_engine):
             error = [400, 'Not Found - missing query', qformat]
             return bad_request(error)
 
-        result = feed_gen(query, engine, count)
+        # first see if we can get the results for the cache
+        engine_and_query = engine + ':' + query
+        result = lookup(engine_and_query)
```
Comment (AnshulMalik) [inline app/server.py]: This would cause issues with number of results

File: app/query_cache.py [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+client = MongoClient(os.environ.get('MONGO_URI', 'mongodb://localhost:27017/'))
+db = client['query-server-v2']
+db = db['queries']  # Automatically delete records that are older than one day
+db.create_index([('createdAt', DESCENDING)], expireAfterSeconds=60 * 60 * 24)
+
+
+def lookup(url):
+    """return search result if the URL is in the db or None on a cache miss."""
+    data = db.find_one({'url': url}) or {}
+    return data.get('links', None)
```
Comment (AnshulMalik) [inline app/query_cache.py]: Will it's result be an array or string? We need to convert this to json or xml so need to be careful.

**Comment by AnshulMalik** [comment]
Why is `help wanted` label stick to this PR? @vaibhavsingh97

File: .travis.yml [hunk 3]
```diff
@@ -8,6 +8,7 @@ install:
 before_script:
   - flake8 . --count --max-complexity=15 --max-line-length=85 --show-source --statistics
 script:
+  - mongod &
```
Comment (AnshulMalik) [inline .travis.yml]: Hey, please have a look at https://docs.travis-ci.com/user/database-setup/#MongoDB

File: app/query_cache.py (hunk 2 above)
Comment (cclauss) [reply app/query_cache.py]: Results at this point are “links” which is a list of dicts.  The encoding and conversation to json, xml happens in the lines that follow.

Comment (AnshulMalik) [reply app/query_cache.py]: Sounds good, I forgot it's mongodb :P

File: app/server.py (hunk 1 above)
Comment (cclauss) [reply app/server.py]: I thought you already deferred that issue.  https://github.com/fossasia/query-server/pull/390#discussion_r157175500

File: .travis.yml (hunk 3 above)
Comment (cclauss) [reply .travis.yml]: Nice!  I have implemented your excellent suggestion.

**Comment by cclauss** [comment]
Ready for review.  @starlord1311 @AnshulMalik @vaibhavsingh97 @bhaveshAn

**Comment by cclauss** [comment]
Can we please remove the “Help Wanted” label and add a “Ready for Review” label in its place?

File: app/server.py (hunk 1 above)
Comment (AnshulMalik) [reply app/server.py]: Cool then

**Comment by cclauss** [comment]
Please provide or update your reviews: image

**Comment by cclauss** [comment]
Done at https://github.com/fossasia/query-server/pull/399/files#diff-73a1e08ff08aa50971c8321d19055e5eR64  Thanks for the reviews and suggestions.

**Comment by vaibhavsingh97** [comment]
@AnshulMalik This PR is ready to merge 🚀 
Should we proceed? You final views please

**Comment by realslimshanky** [comment]
@cclauss please squash your commits into one.

**Comment by cclauss** [comment]
@AnshulMalik

**Comment by cclauss** [comment]
> Please mention about the changes you made in the commit only to the message.

What does that mean?

**Comment by realslimshanky** [comment]
@cclauss you've fixed what was required.

**Comment by cclauss** [comment]
Five positive reviews...  Can we please remove the __needs-review__?  What is the next step in the process?

**Pull request merged**